applicationset/services/scm_provider/gitea.go

```diff
@@ -46,7 +46,16 @@ func NewGiteaProvider(ctx context.Context, owner, token, url string, allBranches
 
 func (g *GiteaProvider) GetBranches(ctx context.Context, repo *Repository) ([]*Repository, error) {
 	if !g.allBranches {
-		branch, _, err := g.client.GetRepoBranch(g.owner, repo.Repository, repo.Branch)
+		branch, status, err := g.client.GetRepoBranch(g.owner, repo.Repository, repo.Branch)
+		if status.StatusCode == 404 {
+			// The gitea API sometimes specifies a non-existing branch as the default. Try to fall back to a reasonable
+			// default.
+			var branches []*gitea.Branch
+			branches, _, err = g.client.ListRepoBranches(g.owner, repo.Repository, gitea.ListRepoBranchesOptions{})
```

Comment:
If the purpose of this function is listing all branches in a repo (my assumption by reading the function name), why not just having this fallback as the main implementation and not have a fallback at all?

Reply:
And wouldn't it fallback if branch was specified by user , but couldn't find

Reply:
@rishabh625 just pushed a change that I think will address this better.  :-)